src/shared/Core/CollectorWriter.cs

```diff
@@ -0,0 +1,172 @@
+                    IsBackground = true
+                };
+
+                _writerThread.Start();
+                // Create a new pipe stream instance using the provided factory
+                NamedPipeClientStream pipe = _createPipeFunc();
+
+                // Specify an instantaneous timeout because we don't want to hold up the
+                // background thread loop if the pipe is not available.
+                pipe.Connect(timeout: 0);
+
+                _pipeClient = pipe;
+            }
+            catch (Exception ex)
+            {
+                RaiseStateChanged(CollectorWriterState.Failing, ex);
+            }
+        }
+
+        public void Dispose()
+        {
+            if (_isDisposed)
+            {
+                return;
+            }
+
+            Stop();
+
+            _pipeClient?.Dispose();
+            _queue.Dispose();
+
+            _pipeClient = null;
+            _writerThread = null;
+
+            _isDisposed = true;
+        }
+
+        public void Stop()
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(CollectorWriter));
+            }
+
+            if (_queue.IsAddingCompleted)
+            {
+                return;
+            }
+
+            // Signal to the queue draining thread that it should drain once more and then terminate.
+            _queue.CompleteAdding();
+            _writerThread.Join();
+
+            Debug.Assert(_queue.IsCompleted, "Message queue should be empty after being stopped");
+        }
+
+        public bool TryEnqueue(string message)
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(CollectorWriter));
+            }
+
+            return _queue.TryAdd(message);
+        }
+
+        private void BackgroundWriterThreadProc()
+        {
+            // Drain the queue of all messages currently in the queue.
+            // TryTake() using an infinite timeout will block until either a message is available (returns true)
+            // or the queue has been marked as completed _and_ is empty (returns false).
+            while (_queue.TryTake(out string message, Timeout.Infinite))
+            {
+                if (message != null)
+                {
+                    WriteMessage(message);
+                }
+            }
+
+            RaiseStateChanged(CollectorWriterState.Stopped, null);
+        }
+
+        private void RaiseStateChanged(CollectorWriterState newState, Exception ex)
+        {
+            State = newState;
+        }
+
+        private void WriteMessage(string message)
+        {
+            try
+            {
+                // If we're in byte/stream transmission mode rather than message mode
+                // we should signal the end of each message with a line-feed (LF) character.
+                if (_pipeClient.TransmissionMode == PipeTransmissionMode.Byte)
```

Comment:
We should always write in message mode, so that the daemon doesn't get fragmented messages.

The message should always have a trailing LF (makes debugging and logging easier).

Reply:
Sounds like on the Unix side, we'll always have UDP in byte mode, so ignore this.  The important thing is to always end the message with a LF.  (The JSON encoder should escape/encode  (JSON-quote) any LFs in the actual data.)